### ISSUE TITLE
Same tables, many databases

### ISSUE THREAD

**Issue by mnpenner** [skeema/skeema#21]
We have kind of a strange setup at my company, so I'm not sure if you'll be able to support this use-case, but if you could, that be fantastic! I was contemplating building a very similar tool before I came across [these slides](https://www.percona.com/live/17/sites/default/files/slides/SkeemaPerconaLive2017.pdf).

We have a requirement that each of our clients has their own database (several actually), but every database has *mostly* the same tables. There's a few differences due to bugs (Skeema could help here!) and because they have some feature differences.

Because there's such a large overlap between these databases, we wouldn't want to have to diff and update every copy of the table individually.

Would it be possible for Skeema to have some concept of sharing a table definition across many databases?

I'm not sure how you would represent this in SQL. For my tool, I was going to rebuild the schema definition in JSON so I'd have some flexibility to tack on extra features like this and also to ease automation if we need to do any mass updates.

**Comment by evanelias** [comment]
Currently Skeema only supports this cleanly if the desired end-state is having the *exact same* set of tables on multiple databases, i.e. traditional sharding schemes. In that case, there are a few ways to map a single dir against multiple database shards.

But if you intentionally want to have some feature differences among the shards, that's a bit trickier! Perhaps you could just solve this using symlinks with relative paths? i.e. a setup like this:
* myinstance/
  * shardA/
    * some_common_table1.sql (real file)
    * some_common_table2.sql (real file)
    * some_unique_table.sql (real file)
  * shardB/
    * some_common_table1.sql --> symlink to ../shardA/some_common_table1.sql
    * some_common_table2.sql --> symlink to ../shardA/some_common_table2.sql
    * some_other_unique_table.sql (real file)

Setting up the symlinks would require some manual tedium, but it's a one-time cost. And then after that, you can just change the canonical copy of a table file and then run `skeema diff`, `skeema push`, etc from the top-level dir (`myinstance` in the example above) to easily manage DDL across multiple databases.

Skeema operations should generally work properly with the symlinks. The only issue I'm aware of is that `skeema pull` can be a bit weird when run against a dir containing symlinks. (It still works properly, it's just that it ends up updating whichever canonical file the symlink points to, which may be undesirable. But it's a conceptual issue / not really clear how else this could or should work.)

In terms of version control, git and many other SCM tools do allow symlinks, although sometimes there are some annoyances involved.

**Comment by evanelias** [comment]
> you won't be able to tell which 'shards' are sharing a definition by looking at `some_common_table1.sql`. You would have to find every symlink that points to it

True, although this is fairly easy to script, for example using `find -L . -samefile shardA/some_common_table1.sql` ... granted, not very elegant or fast, especially if the schema repo is very large.

> and it won't show up in your diffing tools.

Not sure I follow that part -- I'd expect symlinks to work pretty gracefully in terms of diffing.

**Comment by evanelias** [comment]
Quick update on this very old issue. We've kept this open because several variations of it have come up over the years, from a few different users. Some of the simpler cases have been solved by recent features [lax-column-order](https://www.skeema.io/docs/options/#lax-column-order) and [lax-comments](https://www.skeema.io/docs/options/#lax-comments), but the more complex cases haven't been addressed yet.

The [roadmap for Skeema v2](https://www.skeema.io/blog/2025/05/12/skeema-v2-roadmap-rfc/) includes a new automation-friendly subcommand (tentatively called `skeema exec`) which could finally help here, by virtue of allowing STDIN input, instead of strictly operating on .sql files.

So with some custom scripting, it would become possible to arbitrarily transform the set of CREATE TABLE statements depending on what host/database is being pushed to.

Initially this will just be textual SQL, but later on it will also support a JSON format.

Longer-term, another possibility would be support for Go template logic in .sql files, but that would depend on whether enough users need that kind of feature. Most likely, this issue will finally be closed once v2 is released, and then a new issue for template support could be opened if/when there's any demand for it.
